### ISSUE TITLE
wait_for applies delay to each item in with_items

### ISSUE THREAD

**Issue by yourcelf** [ansible/ansible-modules-core#4658]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

`wait_for`
##### ANSIBLE VERSION

```
2.1.1.0
```
##### OS / ENVIRONMENT

Ubuntu 14.0.4.5
##### SUMMARY

Counter-intuitively, the `wait_for` module will apply the `delay` parameter to each item when used `with_items`.  So executing `wait_for` with `delay=60` with 20 items will add 20 minutes total of waiting.
##### STEPS TO REPRODUCE

  For example, the following task (taken from the [ec2 example documentation](https://docs.ansible.com/ansible/ec2_module.html)) will apply an additional 60 second delay before starting each of the items in the loop:

```
    - name: Wait for SSH to come up
      wait_for: host={{ item.public_dns_name }} port=22 delay=60 timeout=320 state=started
      with_items: '{{ec2.instances}}'
```

This is suboptimal behavior for this example -- when launching ec2 instances, it's reasonable to wait some time prior to checking whether they've started.  But since they launch in parallel, it's undesirable to wait an additional 60 seconds between each check.  It'd be more desirable to wait 60 seconds up front, then execute all loop items without further delay.

If delay-for-each-item is intended behavior, it's highly counter-intuitive.  I would argue that this is less useful than applying the delay once prior to execution.  If this behavior is kept as-is, I would suggest improving both the documentation of the `delay` pattern to explain that it is applied to each item in the loop, and changing the example in the ec2 module to instead use the `pause` module, and then executing `wait_for` with no delay.

##### EXPECTED RESULTS

I expected the `delay` parameter would apply once, prior to execution of the loop.
##### ACTUAL RESULTS

The `delay` is applied before each item in the loop.

**Comment by ansibot** [comment]
@gregswift ping, this issue is waiting for your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by resmo** [comment]
notabug

Rather then using with_items it is more practical to apply one task per ec2 instance but with local_action. As an example see this example (using a different cloud) but it pretty much covers your example https://github.com/resmo/ansible-cloudstack-example/blob/master/playbooks/cloud.yml

**Comment by yourcelf** [comment]
I don't believe your example is relevant to my use case -- I'm creating 20 copies of a single vm, with local execution of the play, e.g.:

```
- name: Create load testers
  hosts: localhost
  connection: local
  gather_facts: False
  vars_files:
    - vars/secrets.yml
    - vars/loadtest.yml
  tasks:
    - name: Launch ec2 instances
      ec2:
        aws_access_key: "{{aws_access_key}}"
        aws_secret_key: "{{aws_secret_key}}"
        key_name: "{{ec2_key_name}}"
        instance_type: "{{ec2_instance_type}}"
        image: "{{ec2_image}}"
        region: "{{ec2_region}}"
        group_id: "{{ec2_security_group_id}}"
        instance_tags: "{{ec2_instance_tags}}"
        count: 20
        wait: true
      register: ec2group

    - name: Add instances to host group
      add_host: hostname={{item.public_ip}} groupname={{ec2_host_group_name}}
      with_items: '{{ec2group.instances}}'

    - name: Wait for SSH to come up.
      wait_for: host={{item.public_dns_name}} port=22 delay=60 timeout=320 state=started
      with_items: '{{ec2group.instances}}'
```

So this pretty much has to be done `with_items`.  At any rate, if this behavior of `delay` in `wait_for` is expected, I'll open a separate issue for a docs change for the ec2 example and wait_for.

**Comment by sivel** [comment]
The best way to handle this is to not run your `wait_for` immediately after `add_host`, but instead break into a new play targeting `ec2_host_group_name`, and do your wait_for there.  Similar to:

```
- name: Create load testers
  hosts: localhost
  connection: local
  gather_facts: False
  vars_files:
    - vars/secrets.yml
    - vars/loadtest.yml
  tasks:
    - name: Launch ec2 instances
      ec2:
        aws_access_key: "{{aws_access_key}}"
        aws_secret_key: "{{aws_secret_key}}"
        key_name: "{{ec2_key_name}}"
        instance_type: "{{ec2_instance_type}}"
        image: "{{ec2_image}}"
        region: "{{ec2_region}}"
        group_id: "{{ec2_security_group_id}}"
        instance_tags: "{{ec2_instance_tags}}"
        count: 20
        wait: true
      register: ec2group

    - name: Add instances to host group
      add_host: hostname={{item.public_ip}} groupname={{ec2_host_group_name}}
      with_items: '{{ec2group.instances}}'

- name: configure load testers
  hosts: "{{ ec2_host_group_name }}"
  gather_facts: false
  tasks:
    - name: Wait for SSH to come up.
      local_action: wait_for host={{inventory_hostname}} port=22 delay=60 timeout=320 state=started

    - setup:

    # Do more tasks
```

If you have further questions please stop by IRC or the mailing list:
- IRC: #ansible on irc.freenode.net
- mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by tehmoon** [comment]
Hi @sivel, Thank you for your answer.
It is fine to do that if:
- `ec2_host_group_name`'s length is lower than `--fork`
- You actually use your own machine to perform `wait_for`'s tests.

I don't have access to the rest of the network from where I am. I'm using another SSH server to connect into the infrastructure like this:

```
- wait_for: >
     delay=200
     timeout=320
     state=started
     port={{ hostvars[item.id]["ansible_ssh_port"] }}
     host={{ hostvars[item.id]["ansible_ssh_host"] }}
   with_items: "{{ ec2.instances | default([]) }}"
   when: ( item.state == 'running' )
   delegate_to: "{{ groups['jump_host'][0] }}"
```

I ended up doing this:

```
- command: "sleep 200"
   delegate_to: localhost

 - wait_for: >
     delay=0
     timeout=10
     state=started
     port={{ hostvars[item.id]["ansible_ssh_port"] }}
     host={{ hostvars[item.id]["ansible_ssh_host"] }}
   with_items: "{{ ec2.instances | default([]) }}"
   when: ( item.state == 'running' )
   delegate_to: "{{ groups['jump_host'][0] }}"
```

Also I want to add that I tried `async/poll` without success.

Hope it helps.

**Comment by elnur** [comment]
That's what I ended up with:
```yaml
- hosts: foo
  become: yes
  remote_user: ubuntu
  gather_facts: no

  pre_tasks:
    - name: Wait for SSH
      wait_for:
        host: "{{ inventory_hostname }}"
        port: 22
        state: started
        delay: 60
      delegate_to: localhost
      become: no

    - name: Install Python
      raw: apt install python -y

    - name: Gather facts
      setup:
```
